### PR TITLE
Signup View Model Refactor

### PR DESCRIPTION
**Pull request by ifbarrera**
# 📲 What

Following a great talk at the Swift Functional Conf by @stephencelis on refactoring our view models, I decided to take a crack at this by taking some elements of the refactor from Stephen’s talk, while also attempting to reduce the impact on view model tests. I also wanted to see how it might look without bringing in `ReactiveCocoa` as a dependency (something we should evaluate separately).

# 🛠 How

View model input signal/observer pairs are stored as properties within an `Inputs` struct in the view model. They are immutable, and are created using `Signal.pipe()`. They are publicly accessible, which means observers can be sent values using `observer.send(value: value)` from view controllers as needed. The inputs are fed as an argument into the `func outputs(from inputs: Inputs)` function, which returns a type-aliased `Outputs` tuple. All the logic happens within this function.

With this approach, the tests don’t really change much (no additional boilerplate) which is nice, and the usage in the ViewController is also very similar. In `bindViewModel`, you call the `outputs(from:)` function and then use the outputs as needed within `bindViewModel`.

Most of the benefits of this overall approach are described in [Stephen’s PR](https://github.com/kickstarter/ios-oss/pull/504), but some additional benefits in *this* PR include:
- Doesn’t require Reactive Cocoa dependency
- Don’t need to store reactive properties in the view controller, they stay in the view model
- The tests don’t get any more verbose: can reuse the inputs from the view model in the tests
- gets rid of `MutableProperty`, which saves us from having to initialize them with initial values, plus cuts down on the verboseness of having to call `property.signal`

Drawbacks:
- To refactor the view model tests, one must change all calls to input functions to instead call `send(value: value)`. This is a bit tedious because all the old input functions have different names so a simple search & replace won’t work here.

All in all… this refactor nets approx. -48 lines of code! Not bad.

# 👀 See

Signup works as usual:
tqitputjdn

# ✅ Acceptance criteria

- [ ] Sign-up form works as expected

### CODE REVIEW

File: Library/ViewModels/SignupViewModel.swift [hunk 1]
```diff
@@ -4,108 +4,85 @@ import Prelude
...
+    public let (passwordTextFieldDidReturnSignal, passwordTextFieldDidReturnObserver) =
+      Signal<(), NoError>.pipe()
+    /// Call when the user taps signup.
+    public let (signupButtonPressedSignal, signupButtonPressedObserver) = Signal<(), NoError>.pipe()
+    /// Call when the view did load.
+    public let (viewDidLoadSignal, viewDidLoadObserver) = Signal<(), NoError>.pipe()
+    /// Call when the user toggles weekly newsletter.
+    public let (weeklyNewsletterChangedSignal, weeklyNewsletterChangedObserver) = Signal<Bool, NoError>.pipe()
+
+    init() {}
```
Comment (justinswart) [inline Library/ViewModels/SignupViewModel.swift]: Strange that Swift wouldn't let us get away without this when we use the `pipe()`s!

File: Library/ViewModels/SignupViewModel.swift [hunk 2]
```diff
@@ -4,108 +4,85 @@ import Prelude
...
+    postNotification: Signal<Notification, NoError>,
+    /// Sets whether the name text field is the first responder.
+    nameTextFieldBecomeFirstResponder: Signal<(), NoError>,
+    /// Emits the value for the weekly newsletter.
+    setWeeklyNewsletterState: Signal<Bool, NoError>,
+    /// Emits when a signup error has occurred and a message should be displayed.
+    showError: Signal<String, NoError>
+  )
+
+  public let inputs = Inputs()
```
Comment (justinswart) [inline Library/ViewModels/SignupViewModel.swift]: Why do we have this local `let` again?

Comment (ifbarrera) [reply Library/ViewModels/SignupViewModel.swift]: We store a reference to the inputs in order to send them values (via self.vm.inputs.input.send(value: value)` in both the view controller and the view model tests, but also in order to pass inputs into the `outputs(for inputs:)` function to generate our outputs.

Comment (justinswart) [reply Library/ViewModels/SignupViewModel.swift]: Oh right... why do we need to pass it in as an argument if it's a local var? Couldn't `outputs()` take no arguments and return `self.inputs`?

Comment (ifbarrera) [reply Library/ViewModels/SignupViewModel.swift]: Yep, that could also work.

Comment (justinswart) [reply Library/ViewModels/SignupViewModel.swift]: Cool, yeah I think if there's no real reason to pass it in then that makes sense? It doesn't help us with tests as far as I can tell.

**Comment by haaakon** [comment]
Interesting approach. We at our company love the way you guys set up your `ViewModels`, and i was trying this new refactored way of doing them. Only caveat i found was when configuring `ViewControllers` one has to make sure the outputs function is called on the `ViewController` init, and not on `viewDidLoad`, since there can be input signals firing before a `ViewController` will get its `viewDidLoad` function run. Dunno if thats a problem in your app, but if anyone else tries this new approach its just something to have in mind :)

**Comment by justinswart** [comment]
@haaakon what we’ve done in the past to work around this sort of race condition where an output may emit before a necessary lifecycle event occurs is to combine the output with that event’s input signal. So for example if we need an output to emit _after_ `viewDidLoad` is called then we would create a combined signal of the output we’re interested in and `viewDidLoad`’s input signal.

However am I misunderstanding your concern? Maybe you can share some example code that demonstrates the problem?

**Comment by haaakon** [comment]
We often do 

```
let newViewController = NewViewController.init()
newViewController.inputs.configureData(myData)
present(newViewController)
```
and with this pattern, the `bindViewModel` is called in the `ViewController` `viewDidLoad()`
``` 
override func viewDidLoad() { 
      bindViewModel()
      bindStyles()
      vm.inputs.viewDidLoadObserver.send(value:())
} 

func bindViewModel() {
   let (variable1, variable2) = self.outputs() 
}
```

means 
```
func outputs() -> Outputs {
    let variable1 = Signal.combineLatest(viewDidLoadSignal,  configureDataSignal)    
    }

```
`ConfigureData` is called before this `combineLatest` is set up, since the outputs are generated at `bindViewModel()`. Then the `combineLatest` will "miss" a firing of configureData, if configureData is called immedeately when a viewController is created

**Comment by justinswart** [comment]
@haaakon aah right. Yes that's a good catch, thanks. This may actually be a reason to keep using `MutableProperty<T>` in such cases 🤔 If we set the `value` on a property rather than just send values through pipes then the `MutableProperty<T>` should retain its last value and emit that when it's converted to a `SignalProducer<T, Error>`. I _believe_ you may actually want to use `.producer` and not `.signal` in this case (sorry, I'm not currently able to test it), see this for reference: https://github.com/ReactiveCocoa/ReactiveSwift/issues/620.

We like `pipe()` because it's simpler than using `MutableProperty` in that we don't need an initial `nil` value and then aren't later required have to `skipNil()` - but it does appear that in this case it could be problematic.

@ifbarrera perhaps we should pair on a few more of these edge cases when I'm back 😅

**Comment by haaakon** [comment]
A quick & dirty way of fixing this is doing
```
class MyViewController() : UIViewController {
      let vm = MyViewModel()
      var outputs: MyViewModel.Outputs!
 
    override public init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.outputs = vm.outputs()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.outputs = vm.outputs()
    }
     
    func  bindViewModel() { 
        outputs.myVariable.observeValues { [weak self] value in 
               print(value)
        }
    }
```
That solves the problem for `ViewControllers` at least, but i havent gotten to `UIView` subclasses yet. We also do alot of inputs with `Property<T>` since we "stream" data in our app, so will have to do more research if that still works.

I would hope one does not need to go back to MutableProperty<T> though, its so verbose and i don't like the skipNil() on everything. This new pattern saves time, is easier to understand and cleaner in my opinion, so i will definately follow this PR to see if you have and suggestions!

**Comment by ifbarrera** [comment]
After discussing with the rest of the team, we've decided that the slight reduction boilerplate code for view models with this refactor is probably not worth the inconsistency it will introduce into the codebase. However, we are still investigating ways to ensure that `bindViewModel` is called as early as possible - something we're exploring is using a protocol with a default implementation of an `instantiate()` function that calls `bindViewModel()` immediately after initialization on programmatic view controllers. Combined with a little bit of extra logic in our swizzled `ksr_viewDidLoad()` function to make sure `bindViewModel()` isn't called multiple times, we can ensure that view models are bound as early as possible to prevent possible issues with missed signals in the future.

**Pull request closed without merging**